CDAP-8942 property to enable/disable update schedule in cdap-site

gsps1 opened the pull request:
Add "app.update.schedules" to cdap-site and corresponding changes in AppLifecycle handler and AppRequest.
Missed adding this property changes in last PR.

gsps1 committed:
CDAP-8942 property to enable/disable update schedule in cdap-site

gsps1 commented:
TODO : Will add documentation for this.

awholegunch commented:
Added a Quick Build for this branch, as this will break the docs: https://builds.cask.co/browse/CDAP-DQB332-1

See log output at: https://builds.cask.co/browse/CDAP-DQB332-1/log

awholegunch commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -414,6 +414,14 @@
+    <name>apps.update.schedules</name>
+    <value>true</value>
+    <description>
+      Enable if changes in schedules of an application can be updated on app redeploy.
```
Need to discuss this, but I think this might be better:
```
If true, redeploying an application will update any schedules that currently
exist for the application; if false, existing schedules are maintained 
without change

rohitsinha54 replied:
This property does not affect update from schedule endpoint. It will be good to explicitly say this here

rohitsinha54 replied:
Also say that new schedule will not be created if its set to false

awholegunch replied:
What about:
```
    <description>
      If true, redeploying an application will modify any schedules that currently exist
      for the application; if false, redeploying an application does not create any new
      schedules and existing schedules are neither deleted nor updated. This property only
      affects the redeployment of an application; all related actions or endpoints are
      unaffected.
    </description>

prinam commented:
Build passed: https://builds.cask.co/browse/CDAP-RUT896

rohitsinha54 commented on cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:
```diff
@@ -215,6 +216,8 @@
      */
     public static final String PROGRAM_CONTAINER_DIST_JARS = "program.container.dist.jars";
+
+    public static final String APP_UPDATE_SCHEDULES = "app.update.schedules";
```
@poornachandra suggested we make it "app.deploy.update.schedules"
It will be good to rename the header string X- thing to match this too

rohitsinha54 commented on cdap-proto/src/main/java/co/cask/cdap/proto/artifact/AppRequest.java:
```diff
@@ -41,19 +41,19 @@ public AppRequest(ArtifactSummary artifact) {
 
   public AppRequest(ArtifactSummary artifact, @Nullable T config) {
-    this(artifact, config, null, null, true);
+    this(artifact, config, null, null, null);
```
Don't like the idea of having a boolean as null we can initialize with true or false whatever the default value is here.

gsps1 replied:
we want to differentiate between "not-provided" vs "provided" so have to keep this null here.

rohitsinha54 replied:
But this has a default behavior if not specified the value should represent that in my opinion. Also it's a Boolean so expectation is that it will true/false. Null can cause issues.

rohitsinha54 replied:
I talked to @poornachandra about the null thing and he suggested to keep null if not specified so ignore the null comment earlier.

rohitsinha54 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java:
```diff
@@ -129,6 +130,8 @@
     this.applicationLifecycleService = applicationLifecycleService;
     this.tmpDir = new File(new File(configuration.get(Constants.CFG_LOCAL_DATA_DIR)),
                            configuration.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    this.appUpdateSchedules = configuration.getBoolean(Constants.AppFabric.APP_UPDATE_SCHEDULES,
```
This can go in the ApplicationLifecycleService

rohitsinha54 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java:
```diff
@@ -346,8 +349,11 @@ public void updateApp(HttpRequest request, HttpResponder responder,
+
     try {
-      applicationLifecycleService.updateApp(appId, appRequest, createProgramTerminator());
+      applicationLifecycleService.updateApp(appId, appRequest, createProgramTerminator(), updateSchedules);
```
If we make ApplicationLifecycleService responsible for reading Constants.AppFabric.APP_UPDATE_SCHEDULES and determining the appropriate value of whether to update schedule or not we will not need to pass this from here

gsps1 replied:
thing is deployApp method call in AppLifecycleService expects `updateSchedules`, this deploy app is called from AppLifecycleHttpHandler, so Handler anyway has that default value of update schedules, in order to not pass this update schedule here, there is duplication of getting the default update schedule behavior in handler and lifecycle service, which doesn't provide much advantage. i prefer to pass this value along.

rohitsinha54 replied:
Why does the AppLifecycleHttpHandler need appUpdateSchedules property to be read? In the current code I see it is just passed to the lifecycleService. If you move it to service then you don't need it in handler so there is no duplication as such.

gsps1 replied:
i was referring to the `deployApp` method in LifecycleService which is called directly from the handler.

rohitsinha54 replied:
I see. You can pass in the appRequest.canUpdateSchedules() to the deployApp from the handlers to the lifecycleService and then the 

lifecycleService can read the cdap-site.xml and and do this
```
    boolean updateSchedules =
      appRequest.canUpdateSchedules() == null ? appUpdateSchedules : appRequest.canUpdateSchedules();
```

Which currently we are doing in 2-3 places in handler. It can be done just in the deployApp of the lifecycleService (the main deployApp which deos stuff around line 589)

rohitsinha54 reviewed (changes requested):
Very few minor comments about documentation and one for avoiding null.

gsps1 commented:
@rohitsinha54 
updated the property and added necessary update for "create app" command in CLI to work. 
new build running here : https://builds.cask.co/browse/CDAP-RUT896-2/

replied to other comments.

Also passing `defaultValue` to the canUpdateSchedules is also not an option, as its used in CreateAppCommand in CLI, where we don't know what the default values are.

rohitsinha54 commented:
> Also passing defaultValue to the canUpdateSchedules is also not an option, as its used in CreateAppCommand in CLI, where we don't know what the default values are.

Just make the lifecycleService responsible for reading the property from cdap-site. The canUpdateSchedules can return true or false depending on what is provided and the defaultValue if not provided.

rohitsinha54 commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/app/CreateAppCommand.java:
```diff
@@ -66,17 +67,22 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
     JsonObject config = new JsonObject();
     String ownerPrincipal = null;
+    Boolean updateSchedules = null;
+    PreviewConfig previewConfig = null;
```
👍

rohitsinha54 reviewed (changes requested):
Few comments.

rohitsinha54 commented:
Did  we added docs for app deployment ?

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -126,6 +132,9 @@ and Kerberos principal with which the application is to be deployed (if required
   X-Principal: <Kerberos Principal>
 
+and enable/disable updating schedules of the existing workflows using the header::
+  X-App-Deploy-Update-Schedules: <Update Schedules>
```
insert a blank line before line 136

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -77,6 +78,11 @@ Optionally, you can specify a Kerberos principal with which the application shou
 If a Kerberos principal is specified, then all the streams and datasets created by the
 application will be created with the application's Kerberos principal.
 
+Optionally, you can set/reset the flag "app.deploy.update.schedules". If true,
```
```
Optionally, you can set or reset the flag ``app.deploy.update.schedules``. If true,

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -77,6 +78,11 @@ Optionally, you can specify a Kerberos principal with which the application shou
+Optionally, you can set/reset the flag "app.deploy.update.schedules". If true,
+redeploying an application will modify any schedules that currently exist for the application;
+if false, redeploying an application does not create any new schedules and existing schedules
+are neither deleted nor updated
```
add a trailing period.

awholegunch reviewed (changes requested):
LGTM after comments are addressed.

rohitsinha54 commented:
LGTM

gsps1 commented:
Rebased commits, will merge after the final build https://builds.cask.co/browse/CDAP-RUT896-5
